plugins/crypto/ua_securitypolicy_filestore.c

```diff
@@ -0,0 +1,411 @@
+                                          const UA_ByteString localPrivateKey,
+                                          const UA_String storePath,
+                                          const UA_Logger *logger) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    memset(policy, 0, sizeof(UA_SecurityPolicy));
+
+    SecurityPolicy_FilestoreContext *pc = (SecurityPolicy_FilestoreContext *)
+        UA_calloc(1, sizeof(SecurityPolicy_FilestoreContext));
+    if(!pc)
+        return UA_STATUSCODE_BADOUTOFMEMORY;
+
+    UA_String_copy(&storePath, &pc->storePath);
+    policy->policyContext = (void *)pc;
+
+    retval = UA_SecurityPolicy_Basic128Rsa15(&pc->policy, localCertificate,
+                                             localPrivateKey, logger);
+    if(retval != UA_STATUSCODE_GOOD)
+        return retval;
+
+    init_sp_filestore(policy);
+
+    return retval;
+}
+
+UA_StatusCode
+UA_SecurityPolicy_Basic256_Filestore(UA_SecurityPolicy *policy,
+                                     const UA_ByteString localCertificate,
+                                     const UA_ByteString localPrivateKey,
+                                     const UA_String storePath,
+                                     const UA_Logger *logger) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    memset(policy, 0, sizeof(UA_SecurityPolicy));
+
+    SecurityPolicy_FilestoreContext *pc = (SecurityPolicy_FilestoreContext *)
+        UA_calloc(1, sizeof(SecurityPolicy_FilestoreContext));
+    if(!pc)
+        return UA_STATUSCODE_BADOUTOFMEMORY;
+
+    UA_String_copy(&storePath, &pc->storePath);
+    policy->policyContext = (void *)pc;
+
+    retval = UA_SecurityPolicy_Basic256(&pc->policy, localCertificate,
+                                        localPrivateKey, logger);
+    if(retval != UA_STATUSCODE_GOOD)
+        return retval;
+
+    init_sp_filestore(policy);
+
+    return retval;
+}
+
+UA_StatusCode
+UA_SecurityPolicy_Basic256Sha256_Filestore(UA_SecurityPolicy *policy,
+                                           const UA_ByteString localCertificate,
+                                           const UA_ByteString localPrivateKey,
+                                           const UA_String storePath,
+                                           const UA_Logger *logger) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    memset(policy, 0, sizeof(UA_SecurityPolicy));
+
+    SecurityPolicy_FilestoreContext *pc = (SecurityPolicy_FilestoreContext *)
+        UA_calloc(1, sizeof(SecurityPolicy_FilestoreContext));
+    if(!pc)
+        return UA_STATUSCODE_BADOUTOFMEMORY;
+
+    UA_String_copy(&storePath, &pc->storePath);
+    policy->policyContext = (void *)pc;
+
+    retval = UA_SecurityPolicy_Basic256Sha256(&pc->policy, localCertificate,
+                                              localPrivateKey, logger);
+    if(retval != UA_STATUSCODE_GOOD)
+        return retval;
+
+    init_sp_filestore(policy);
```

Comment:
I think I get the idea behind this approach: You basically create a second SecurityPolicy inside the SecurityPolicy_FilestoreContext, initialize this "inside" policy with UA_SecurityPolicy_Basic256Sha256 and then copy the functions pointers from the "inside" policy to the "outside" policy, replacing those functions that need to do additional work (storing an updated Application Instance Certificate in the certificate store).

However, this does not work (at least for me), as the functions behind the copied function pointers (e.g. "asymmetricModule_compareCertificateThumbprint_sp_basic256sha256") expect the policyContext to be of type "Basic256Sha256_PolicyContext", but get a "SecurityPolicy_FilestoreContext" instead.

To make this work, you will IMHO either have to provide "trampoline" functions for _each and every_ function registered in the SecurityPolicy that accesses the policyContext, which will then call the actual function with the "inside" policy (as is done by "updateCertificateAndPrivateKey_sp_filestore"), or add something like an additional "storeContext" in the UA_SecurityPolicy structure. Seeing that the "transactions" feature waiting to be included does not use the additional "inside" SecurityPolicy for storing the new certificate and private key, this would allow to get rid of the SecurityPolicy structure duplications, as far as I see.